Add "Compare View" Column to Catalog for Outputting Comparison Views

danieldresser-ie opened the pull request:
We've discussed this functionality a fair bit - I think the implementation makes a fair bit of sense.  I'm less confident that it feels good to use, but I think it does do basically what we've talked about.  The lack of interactions like highlighting definitely makes it less inviting ... there's also currently something going wrong with the icons on the left of the catalog ... but should be ready for you to take a look.

danieldresser-ie commented:
Uploaded the current mess in whatever state it was in yesterday.  I think this probably is a pretty decent direction for where to go with things.

johnhaddon commented:
I like the direction the Catalogue is going in, so let's get that polished up and over the line before tackling the thornier viewer integration questions. With that in mind, I have a few thoughts :

- I think CatalogueSelect should support this new functionality, so it isn't just available in the Viewer. With that in mind :
   - I wonder if "compare #" is the right description here. What about "bookmark #"? Or "output #"? Or "tag #"? Something else along those lines?
   - I almost wonder if the primary `Catalogue.imageIndex` should just be "tag 0" or "tag 1", so we've only got one concept rather than two (using "tag" for exposition only, I'm not attached to it particularly"). Eliminating `Catalogue.imageIndex` would also be nice because then everything is encapsulated under `Catalogue.images`, and there's only one plug to promote.
   - I wonder if having two context variables is necessary - it creates ambiguity as to which takes precedence. Perhaps something like `context.set( "catalogue:imageName", "tag:1" )` would be better?
- Could do with some unit tests, including :
    - Getting the expected image from CatalogueSelect
    - Maintaining the exclusivity of particular tag values
    - All of the above while the `images` plug is promoted
- Would be good to use the same filled style for the heading (with the # in) as the cell values.

danieldresser-ie commented:
I think I've addressed all the feedback we've discussed, other than still needing to add tests ( wasn't quite as productive as I hoped today ).  Other than needing tests, I think it's in an OK state to review, though.

johnhaddon commented on resources/graphics.py:
```diff
@@ -161,6 +161,11 @@
 				'compareModeSideBySide',
 				'wipeDisabled',
 				'wipeEnabled',
+				'catalogueOutputHeader',
+				'nodeSetCatalogueOutput1',
+				'nodeSetCatalogueOutput2',
+				'nodeSetCatalogueOutput3',
+				'nodeSetCatalogueOutput4',
```
These are really soft on my monitor, especially the header one. When making icons this small, it makes a huge difference whether or not the objects are aligned to the pixel grid. In this case the border is 1.5 pixels, so naturally soft, and the # mark is completely unaligned. I experimented with adjusting to the grid in https://github.com/johnhaddon/gaffer/commit/04b44988d579bffd3ec8bc735500c777f4339fbc - still not great, but a bit better I think? I wonder if we'd be better off using a square than a circle?

Also, the icons are in the ImageView section - they'd be better off in the Catalogue section.

johnhaddon replied:
Oh, and we should drop the `nodeSet` prefix.

johnhaddon commented on src/GafferImage/Catalogue.cpp:
```diff
@@ -1131,14 +1148,37 @@ void Catalogue::imageReceived( Gaffer::Plug *plug )
+		{
+			if( image != currentImage && image->outputIndexPlug()->getValue() == newIndex )
+			{
+				image->outputIndexPlug()->setValue( 0 );
```
This is going to break when you add unit tests for promoted plugs - it needs to set `image->outputIndexPlug()->source()`.

johnhaddon commented on src/GafferImage/Catalogue.cpp:
```diff
@@ -1131,14 +1148,37 @@ void Catalogue::imageReceived( Gaffer::Plug *plug )
+
+void Catalogue::plugSet( const Plug *plug )
+{
+	// Enforce that only one image may have a particular output index
```
It'd be nice to document the rationale for having `Image.outputIndex` rather than `Catalogue.outputIndices` - it being easier to enforce uniqueness than it is to sync indices during deletion/reordering. Could also add a todo to handle selection the same way?

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -1096,6 +1119,31 @@ def __keyPress( self, imageListing, keyEvent ) :
+		return True
+
+	def __setComparisonImage( self, index, image, *unused ) :
+		image["outputIndex"].setValue( index )
```
I guess this is deliberately not wrapped in an UndoScope? Because we think the output index stuff should feel like the selection, and that isn't undoable either?

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -1116,20 +1164,38 @@ def __columnContextMenuDefinition( self ) :
+		menu = IECore.MenuDefinition()
+		for i in range( 1, 5 ) :
+
+			menu.append( "/Output Index %i" % i, {
+				"checkBox" : image["outputIndex"].getValue() == i,
+				"command" : functools.partial( Gaffer.WeakMethod( self.__setComparisonImage ), i, image  ),
+			} )
```
I feel like this might read a bit better if we omitted the repeated "Output Index" on every line, and instead put that in the menu title? So the first thing you read on each line is the part that makes it unique?

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -1116,20 +1164,38 @@ def __columnContextMenuDefinition( self ) :
+		elif isinstance( self.__pathListing.columnAt( mousePosition ), OutputIndexColumn ):
+			image = _imageFromName( self.getPlug().node(), self.__pathListing.pathAt( mousePosition )[0] )
+			if image:
+				self.__popupMenu = GafferUI.Menu( self.__outputIndexContextMenuDefinition( image ), "Set As Output Index" )
```
`title = "Output Index"`. `"Set as Output Index"` is a positional argument getting swallowed by `_qtParent`.

johnhaddon commented on python/GafferImageUI/CatalogueSelectUI.py:
```diff
@@ -51,11 +51,17 @@ def __imageNames( plug ) :
 
-	return IECore.StringVectorData( [ "Selected" ] + __imageNames( plug ) )
+	return IECore.StringVectorData(
+		[ "Selected", "Output 1", "Output 2", "Output 3", "Output 4" ] +
+		[ "Image Name/" + i for i in __imageNames( plug ) ]
```
This feels a bit better to me if the outputs are in a submenu too, and we drop the "Name" part from the "Image Name" submenu.

![image](https://user-images.githubusercontent.com/1133871/192517208-62bfe988-a8db-408a-ad5b-73cd63847534.png)

danieldresser-ie commented:
Addressed feedback, added tests, switched ImageView comparison to operate on catalog outputs.

johnhaddon commented on src/GafferImage/Catalogue.cpp:
```diff
@@ -1148,57 +1214,100 @@ void Catalogue::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outp
+		}
+		else
+		{
+			result = "Catalogue Unknown Image: \"" + imageName + "\"";
```
Maybe `Catalogue : Unknown Image "<imageName>"`? "Catalogue Unknown Image" without the separator doesn't parse very well in my brain...

johnhaddon commented on src/GafferImage/Catalogue.cpp:
```diff
@@ -1148,57 +1214,100 @@ void Catalogue::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outp
+		std::string result;
+		if( outputIndex > 0 )
+		{
+			result = "Catalogue Output " + std::to_string( outputIndex );
```
Maybe `Catalogue : Output <N>` or `Catalogue : Unassigned Output <N>`, to match the unknown image message a bit better?

johnhaddon commented on src/GafferImage/Catalogue.cpp:
```diff
@@ -1148,57 +1214,100 @@ void Catalogue::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outp
 			}
 			childIndex++;
 		}
-		if( index < 0 )
-		{
-			throw IECore::Exception( "Unknown image name \"" + imageName + "\"." );
-		}
```
I think we need to update the CatalogueSelect tests for this change :

```
Error: FAIL: testSelection (GafferImageTest.CatalogueSelectTest.CatalogueSelectTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/gaffer/gaffer/build/python/GafferImageTest/CatalogueSelectTest.py", line 90, in testSelection
    GafferImage.ImageAlgo.image( catalogueSelect["out"] )
AssertionError: Exception not raised
```

johnhaddon reviewed (changes requested):
Thanks Daniel - this really has turned out to be pleasingly simple after some of the alternatives we considered. In addition to the small inline comments, I have a couple more :

- Now that the `compare` section in the ImageView header is so nice and compact, would it make sense to keep all three icons visible all the time, but grey out the last two when comparison is inactive? I find the jump caused by the size change to be slightly jarring, and I think we can probably justify the real estate now?
- The catalogue icon images are still in the ImageView section of the graphics, rather than the catalogue section. I think it probably makes sense to move them?

Other than that, let's get this shipped!

johnhaddon commented:
Oh, and we need to update Changes.md.

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -1096,6 +1119,32 @@ def __keyPress( self, imageListing, keyEvent ) :
+		if not isinstance( cellColumn, OutputIndexColumn ):
+			return False
+
+		if event.button == event.Buttons.Left and event.modifiers == event.Modifiers.None_ :
```
Hmm. Maybe scratch that. I'm not sure how we can get that into the "hover state" scheme for giving pre-click feedback in PathColumn.

danieldresser-ie commented:
OK, I think I hit all feedback.  Main things that need double checking are probably the spreadsheet tooltips ( I wasn't quite sure what to write ), and the new look of the UI ( The stylesheet hack to give consistency between MenuButton and gafferThinButton, and the darkened icon for "compareModeNone.png", since the grey icon looked disabled )

johnhaddon commented:
Thanks! I've got a couple of thoughts on the stylesheet stuff, but I'll present those in a separate PR to see what you think. Merged!
